pkg/controllers/vdb/sandboxsubcluster_reconciler.go

```diff
@@ -0,0 +1,184 @@
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// SandboxSubclusterReconciler will add subclusters to sandboxes
+type SandboxSubclusterReconciler struct {
+	VRec        *VerticaDBReconciler
+	Log         logr.Logger
+	Vdb         *vapi.VerticaDB // Vdb is the CRD we are acting on
+	PFacts      *PodFacts
+	InitiatorIP string // The IP of the pod that we run vclusterOps from
+	Dispatcher  vadmin.Dispatcher
+	client.Client
+}
+
+// MakeSandboxSubclusterReconciler will build a SandboxSubclusterReconciler object
+func MakeSandboxSubclusterReconciler(vdbrecon *VerticaDBReconciler, log logr.Logger, vdb *vapi.VerticaDB,
+	pfacts *PodFacts, dispatcher vadmin.Dispatcher, cli client.Client) controllers.ReconcileActor {
+	return &SandboxSubclusterReconciler{
+		VRec:       vdbrecon,
+		Log:        log.WithName("SandboxSubclusterReconciler"),
+		Vdb:        vdb,
+		PFacts:     pfacts,
+		Dispatcher: dispatcher,
+		Client:     cli,
+	}
+}
+
+// Reconcile will add subclusters to sandboxes if we found any qualified subclusters
+func (s *SandboxSubclusterReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	// no-op for ScheduleOnly init policy or enterprise db or no sandboxes
+	if s.Vdb.Spec.InitPolicy == vapi.CommunalInitPolicyScheduleOnly ||
+		!s.Vdb.IsEON() || len(s.Vdb.Spec.Sandboxes) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// We need to collect pod facts for finding qualified subclusters
+	if err := s.PFacts.Collect(ctx, s.Vdb); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return s.sandboxSubclusters(ctx)
+}
+
+// sandboxSubclusters will add subclusters to their sandboxes defined in the vdb
+func (s *SandboxSubclusterReconciler) sandboxSubclusters(ctx context.Context) (ctrl.Result, error) {
+	// find qualified subclusters with their sandboxes
+	scSbMap, err := s.fetchSubclustersWithSandboxes()
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if len(scSbMap) == 0 {
+		s.Log.Info("No subclusters need to be sandboxed")
+		return ctrl.Result{}, nil
+	}
+
+	// find an initiator to call vclusterOps
+	foundInitiator := false
+	for _, v := range s.PFacts.Detail {
+		// use an UP primary node in main cluster as the initiator
+		if !foundInitiator && v.sandbox == "" && v.isPrimary && v.upNode {
+			s.InitiatorIP = v.podIP
+			foundInitiator = true
+		}
+	}
+	if !foundInitiator {
+		return ctrl.Result{}, fmt.Errorf("cannot find an UP node in main cluster to execute sandbox operation")
+	}
+
+	// call sandbox API in vclusterOps, then update sandbox status in vdb
+	succeedSbScMap := make(map[string][]string)
+	for sc, sb := range scSbMap {
+		err := s.sandboxSubcluster(ctx, sc, sb)
+		if err != nil {
+			// when one subcluster failed to be sandboxed, update sandbox status and return error
+			return ctrl.Result{}, errors.Join(err, s.updateSandboxStatus(ctx, succeedSbScMap))
+		} else {
+			succeedSbScMap[sb] = append(succeedSbScMap[sb], sc)
+		}
+	}
```

Comment:
You can move this to a separate function that returns the map  and an error, and if the map is non-empty you need to invalidate the podfacts(`s.PFacts.Invalidate()`) because some pods have been sandboxed.